### PR TITLE
Improving getting started materials

### PR DESCRIPTION
**Pull request by cjnolet**
Adding documentation for each index type. This decouples the major details in the docs from the individual language API docs to present users with a better overal experience. The major topcs in the new index-level docs are meant to be more exhaustive than the API docs in providing

1. Description of each index type, when it's useful and shortcomings
2. Links to any relevant research material to learn more foundational info
3. Hyper-parameters
4. Links to API docs for each supported language
5. Links to example projects and notebooks
6. Formulas for rough estimates of memory footprint
7. How to use in cuvs-bench, along with some rough benchmarks on different hardware

### CODE REVIEW

File: docs/source/indexes/ivfflat.rst [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+Tuning Considerations
+---------------------
+
+Since IVF methods use clustering to establish spatial locality and partition data points into individual lists, there's an inherent
+assumption that the number of lists, and thus the max size of the data in the index is known up front. For some use-cases, this
+might not matter. For example, most vector databases build many smaller physical approximate nearest neighbors indexes, each from
+fixed-size or maximum-sized immutable segments and so the number of lists can be tuned based on the number of vectors in the indexes.
+
+Empirically, we've found :math:`\sqrt{n_index_vectors}` to be a good starting point for the :math:`n_lists` hyper-parameter. Remember, having more
```
Comment (lowener) [inline docs/source/indexes/ivfflat.rst:80]: The formatting can be weird if the underscores are not escaped.
```suggestion
Empirically, we've found :math:`\sqrt{n\_index\_vectors}` to be a good starting point for the :math:`n\_lists` hyper-parameter. Remember, having more
```

File: docs/source/indexes/ivfflat.rst [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+lists means less points to search within each list, but it could also mean more :math:`n_probes` are needed at search time to reach an acceptable
```
Comment (lowener) [inline docs/source/indexes/ivfflat.rst:81]:
```suggestion
lists means less points to search within each list, but it could also mean more :math:`n\_probes` are needed at search time to reach an acceptable
```

File: docs/source/cpp_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+Step 1: Starting off with cuVS
+==============================
+
+CUDA Development?
+-----------------
+
+If you are reading this tuturial then you probably know about CUDA and its relationship to general-purpose GPU computing (GPGPU). You probably also know about Nvidia GPUs but might not necessarily be familiar with the programming model nor GPU computing. The good news is that extensive knowledge of CUDA and GPUs are not needed in order to get started with or build applications with RAFT. RAFT hides away most of the complexities behind simple single-threaded stateless functions that are inherently asynchronous, meaning the result of a computation isn't necessarily read to be used when the function executes and control is given back to the user. The functions are, however, allowed to be chained together in a sequence of calls that don't need to wait for subsequent computations to complete in order to continue execution. In fact, the only time you need to wait for the computation to complete is when you are ready to use the result.
```
Comment (lowener) [inline docs/source/cpp_tutorial.rst:44]: This feels like how to get started with RAFT to build a GPU algorithm, not how to get started with cuVS to do vector search?

File: docs/source/cpp_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+Step 3: Using brute-force indexes
+=================================
+
+Build brute-force index
+-----------------------
+
+Consider the `(10k, 10k)` shaped random matrix we generated in the previous step. We want to be able to find the k-nearest neighbors for all points of the matrix, or what we refer to as the all-neighbors graph, which means finding the neighbors of all data points within the same matrix.
+.. code-block:: c++
+
+   #include <raft/neighbors/brute_force.cuh>
```
Comment (lowener) [inline docs/source/cpp_tutorial.rst:145]: In this file modify RAFT to cuVS, in the code examples and in the documentation content.

File: docs/source/cuvs_bench/index.rst [hunk 5]
```diff
@@ -0,0 +1,702 @@
...
+                            path to YAML configuration file or directory for algorithms Any run groups found in the specified file/directory will automatically override groups of the same name
+                            present in the default configurations, including `base` (default: None)
+      --dataset DATASET     name of dataset (default: glove-100-inner)
+      --dataset-path DATASET_PATH
+                            path to dataset folder, by default will look in RAPIDS_DATASET_ROOT_DIR if defined, otherwise a datasets subdirectory from the calling directory (default:
+                            os.getcwd()/datasets/)
+      --build
+      --search
+      --algorithms ALGORITHMS
+                            run only comma separated list of named algorithms. If parameters `groups` and `algo-groups are both undefined, then group `base` is run by default (default: None)
```
Comment (lowener) [inline docs/source/cuvs_bench/index.rst:132]:
```suggestion
                            run only comma separated list of named algorithms. If parameters `groups` and `algo-groups` are both undefined, then group `base` is run by default (default: None)
```

File: docs/source/cpp_tutorial.rst (hunk 3 above)
Comment (cjnolet) [reply docs/source/cpp_tutorial.rst:44]: No, this is eventually going to be a guide on how to use cuVS APIs end-to-end. It's not being linked into the docs yet so it's not even reachable atm. Ran out of time.

Comment (cjnolet) [reply docs/source/cpp_tutorial.rst:44]: I don't want to remove it only because it took time to convert it from markdown to rust, but I'm hoping to revise it and link it in sometime during 24.12 (as time permits).

**Comment by cjnolet** [comment]
Moved the cpp tutorial over to a differerent branch for now: https://github.com/rapidsai/cuvs/tree/doc-2412-cpp_tutorial

File: docs/source/choosing_and_configuring_indexes.rst [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+Unfortunately, for large datasets, doing a hyper-parameter optimization on the whole dataset is not always feasible. It is possible, however, to perform a hyper-parameter optimization on the smaller subsets and find reasonably acceptable parameters that should generalize fairly well to the entire dataset. Generally this hyper-parameter optimization will require computing a ground truth on the subset with an exact method like brute-force and then using it to evaluate several searches on randomly sampled vectors.
+
+Full hyper-parameter optimization may also not always be necessary- for example, once you have built a ground truth dataset on a subset, many times you can start by building an index with the default build parameters and then playing around with different search parameters until you get the desired quality and search performance.  For massive indexes that might be multiple terabytes, you could also take this subsampling of, say, 10M vectors, train an index and then tune the search parameters from there. While there might be a small margin of error, the chosen build/search parameters should generalize fairly well for the databases that build locally partitioned indexes.
```
Comment (dantegd) [inline docs/source/choosing_and_configuring_indexes.rst:67]: Thinking from a user perspective, a future small blog post with examples about this might be quite nice to have.

File: README.md [hunk 7]
```diff
@@ -32,32 +29,32 @@ cuVS contains state-of-the-art implementations of several algorithms for running
...
+| `cuvs` | `libcuvs` |
 
 ### Stable release
 
-It is recommended to use [mamba](https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html) to install the desired packages. The following command will install the Python package. You can substitute `cuvs` for any of the packages in the table above:
+It is recommended to use [mamba](https://conda.github.io/conda-libmamba-solver/user-guide/) to install the desired packages. The following command will install the Python package. You can substitute `cuvs` for any of the packages in the table above:
 
 ```bash
-mamba install -c conda-forge -c nvidia -c rapidsai cuvs
+conda install -c conda-forge -c nvidia -c rapidsai cuvs
```
Comment (benfred) [inline README.md:43]: should we keep the command as `mamba` since thats our recommendation to use in the above paragraph?
```suggestion
mamba install -c conda-forge -c nvidia -c rapidsai cuvs
```

File: README.md [hunk 8]
```diff
@@ -32,32 +29,32 @@ cuVS contains state-of-the-art implementations of several algorithms for running
...
 ```
 
 ### Nightlies
 If installing a version that has not yet been released, the `rapidsai` channel can be replaced with `rapidsai-nightly`:
 
 ```bash
-mamba install -c conda-forge -c nvidia -c rapidsai-nightly cuvs=24.10
+conda install -c conda-forge -c nvidia -c rapidsai-nightly cuvs=24.10
```
Comment (benfred) [inline README.md:50]: (?)

```suggestion
mamba install -c conda-forge -c nvidia -c rapidsai-nightly cuvs=24.10
```

File: docs/source/cuvs_bench/build.rst [hunk 9]
```diff
@@ -0,0 +1,59 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Build cuVS Bench From Source
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Dependencies
+============
+
+CUDA 11 and a GPU with Pascal architecture or later are required to run the benchmarks.
```
Comment (benfred) [inline docs/source/cuvs_bench/build.rst:8]: Do we still support pascal gpus ? https://docs.rapids.ai/notices/rsn0034/

File: docs/source/cuvs_bench/index.rst [hunk 10]
```diff
@@ -0,0 +1,702 @@
...
+.. code-block:: bash
+
+   mamba create --name raft_ann_benchmarks
+   conda activate raft_ann_benchmarks
+
+   # to install GPU package:
+   mamba install -c rapidsai -c conda-forge -c nvidia raft-ann-bench=<rapids_version> cuda-version=11.8*
+
+   # to install CPU package for usage in CPU-only systems:
+   mamba install -c rapidsai -c conda-forge  raft-ann-bench-cpu
```
Comment (benfred) [inline docs/source/cuvs_bench/index.rst:34]: btw - feel like we should be consistent about either using conda or mamba (so if we don't change the `conda install` instructions, we should change this)

File: docs/source/getting_started.rst [hunk 11]
```diff
@@ -1,12 +1,120 @@
...
+
+The primary goal of cuVS is to enable speed, scale, and flexibility (in that order)- and one of the important value propositions is to enhance existing software deployments with extensible GPU capabilities to improve pain points while not interrupting parts of the system that work well today with CPU.
+
+
+Using cuVS APIs
 ===============
 
-This guide provides an initial starting point of the basic concepts and using the various APIs in the cuVS software development kit.
+cuVS is a C++ library its core, which is wrapped with a C library and exposed further through various different languages. cuVS currently provides APIs and documentation for :doc:`C <c_api>`, :doc:`C++ <cpp_api>`, :doc:`Python <python_api>`, and :doc:`Rust <rust_api/index>` with more languages in the works. our :doc:`API basics <api_basics>` provides some background and context about the important paradigms and vocabulary types you'll encounter when working with cuVS types.
```
Comment (benfred) [inline docs/source/getting_started.rst:69]: `cuVS is a C++ library its core` doesn't sound quite right to me - maybe should be `cuVS is a C++ library at its core` instead?

**Review by dantegd — COMMENTED** [review]
Overall things look fantastic! Just very few minor comments

File: docs/source/tuning_guide.rst [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+#. Uniformly sample some number of vectors for a test set and do this again for an evaluation set. 1-10% of the vectors in the training set.
+
+#. Use the test set to compute ground truth on the vectors from prior step against all vectors in the training set.
+
+#. Start the HPO tuning process for the training set, using the test vectors for the query set. It's important to make sure your HPO is multi-objective and optimizes for: a) low build time, b) high throughput or low latency sarch (depending on needs), and c) acceptable recall.
+
+#. Use the evaluation dataset to test that the optimal hyper-parameters generalize to unseen points that were not used in the optimization process.
+
+#. Optionally, the above steps multiple times on different uniform sub-samplings. Optimal parameters can then be combined over the multiple monte-optimization iterations. For example, many hyper-parameters can simply be averaged but care might need to be taken for other parameters.
```
Comment (benfred) [inline docs/source/tuning_guide.rst:49]: maybe `monte-carlo optimization` instead of `monte-optimization` ?

**Review by benfred — APPROVED** [review]
amazing work on this - this is a great change.

I left some minor comments, but pre-approving since none of them are blocking

File: docs/source/choosing_and_configuring_indexes.rst (hunk 6 above)
Comment (cjnolet) [reply docs/source/choosing_and_configuring_indexes.rst:67]: Yeah, I like that idea. I had originally planned this (and the othe guides) as a series of blogs but our blog publishing process has gotten so complex that I opted to add them to docs for now. Would still love to do a blog in the future, though.

Comment (cjnolet) [reply docs/source/choosing_and_configuring_indexes.rst:67]: We are also working to get notebooks and scripts together that demonstrate this whole process end to end on a vector database.

File: README.md (hunk 7 above)
Comment (cjnolet) [reply README.md:43]: Actually, we got dinged on this in our VDR because `mamba` apparenlty only works up to Python 3.11, while conda uses the `libmamba` backend direclty. That's one reason I linked to the instructions for enabling the libmamba backend within conda and changed this. Apparently it was too confusing installing mamba for the VDR team whereas conda at least works out of the box.

Comment (cjnolet) [reply README.md:43]: Let me know if there's a better way to word the suggestion to use mamba, or a better set of instructions we should be linking to.

[the remark this answers is not in the document]
Comment (cjnolet) [reply README.md:43]: The only reason I didn't is to try and minimize the number of places we have this info, since it adds to the challenge of keeping everything up to day. I think we should definiely mention pip, though, and link to the build guide. I considered doing this for conda too. What do you think? Or as a user, would you rather see both pip and conda right in the readme?

File: docs/source/cuvs_bench/index.rst (hunk 10 above)
Comment (cjnolet) [reply docs/source/cuvs_bench/index.rst:34]: I agree, that was an accident- I missed that one.

File: docs/source/tuning_guide.rst (hunk 12 above)
Comment (cjnolet) [reply docs/source/tuning_guide.rst:49]: Haha, that was me speed typing to get it done in time for code freeze. Great catch!

File: docs/source/cuvs_bench/build.rst (hunk 9 above)
Comment (cjnolet) [reply docs/source/cuvs_bench/build.rst:8]: Nope, fixed.
